app/src/main/java/com/epfl/drawyourpath/pathDrawing/PathDrawingEndFragment.kt

```diff
@@ -27,36 +30,55 @@ class PathDrawingEndFragment(private val run: Run? = null) : Fragment(R.layout.f
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+
+        // get the ml result
+        val result = Utils.getRunRecognition(pathDrawingModel.getRun()).thenApplyAsync {
+            it.candidates[0]
+        }
+
+        val runToAdd = pathDrawingModel.getRun()
+
+        val cancel = AtomicBoolean(false)
+
         // display the detail final performance of the user
         setupPerformancePreview()
 
         // display the preview of the path
         setupPathPreview()
 
-        // display the form recognized by the ML model
+        // display the shape recognized by the ML model (or loading message before the model is downloaded)
         setupPathDescription()
+        result.thenApplyAsync {
+            if (!cancel.get()) {
+                setupPathDescription(it)
+            }
+        }
 
         // return back to the menu add and save the path back clicking and the back to menu button
         backMenuButton = view.findViewById(R.id.path_drawing_end_back_menu_button)
         backMenuButton.setOnClickListener {
-            userCached.addNewRun(pathDrawingModel.getRun()).thenApplyAsync {
-                returnBackToMenu()
+            cancel.set(true)
+            result.thenApplyAsync {
+                Run(runToAdd.getPath(), runToAdd.getStartTime(), runToAdd.getDuration(), runToAdd.getEndTime(), it.text, it.score?.toDouble() ?: 0.0)
+            }.thenComposeAsync {
```

Comment:
I think there is a problem with this transition:
Imagine that the calculation of the score and shape recognition is not finished and you click on the back to menu button. Then the run score will be zero (instead of the score predicted, when can be calculated after the nback to menu transition).

Reply:
It is something that I have tested and that works thanks to the fact that we have a single activity. It will put the run in the cache and in the database even if we are back on the main fragment.

Reply:
Ok thank you for the clarification. If you have tested then it should work. Thank you